sdk/core/switcher/entry.S

```diff
@@ -779,3 +779,77 @@ exception_entry_asm:
 #endif
 #endif // CONFIG_NO_SWITCHER_SAFETY
 	cret
+
+// Returns whether the trusted stack has space for N more calls.
+	.section .text, "ax", @progbits
+	.p2align 2
+	.type __Z23trusted_stack_has_spacei,@function
+__Z23trusted_stack_has_spacei:
+	li                 a2, TrustedStackFrame_size
+	mul                a2, a0, a2
+	cspecialr          ca0, mtdc
+	clhu               a1, TrustedStack_offset_frameoffset(ca0)
+	cgetlen            a0, ca0
+	sub                a0, a0, a1
+	sltu               a0, a2, a0
+	cret
+
+// Returns the size of the trusted stack.
+	.section .text, "ax", @progbits
+	.p2align 2
+	.type __Z18trusted_stack_sizev,@function
+__Z18trusted_stack_sizev:
+	cspecialr          ca0, mtdc
+	cgetlen            a0, ca0
+	j                  size_to_trusted_stack_frames
+
+// Returns the current trusted stack index.
+	.section .text, "ax", @progbits
+	.p2align 2
+	.type __Z19trusted_stack_indexv,@function
+__Z19trusted_stack_indexv:
+	cspecialr          ca0, mtdc
+	clhu               a0, TrustedStack_offset_frameoffset(ca0)
+
+// Takes a size in bytes and of a trusted stack and returns the number of
+// trusted stack frames that this contains.
+size_to_trusted_stack_frames:
+	addi               a0, a0, -(TrustedStack_offset_frames+TrustedStackFrame_size)
+#ifdef IBEX
+.if TrustedStackFrame_size != 24
+	.err Trusted stack size has changed, assembly must be updated
+.endif
+	// Reciprocal division generated by clang.
+	// On Ibex, mulh is 3 cycles, div is 15, so this should be faster.
+	lui                a1, 174763
+	addi               a1, a1, -1365
+	mulh               a0, a0, a1
+	srli               a1, a0, 31
+	srai               a0, a0, 2
+	add                a0, a0, a1
+#else
+	addi               a1, zero, TrustedStackFrame_size
+	div                a0, a0, a1
+#endif
+	cret
+
+
+.macro export function, prefix=__library_export_libcalls
```

Comment:
Isn't this just a specialized https://github.com/microsoft/cheriot-rtos/blob/8380a1c2806e5c343f70b2f531305fa43110f4f2/sdk/include/compartment-macros-asm.S#L42

Even if you keep it as is, I think the `.section compartment_export_table` wants to be `.section compartment_exports`?

Reply:
Kind of.  Except that it needs to go in a different section because we're not linking this.

Reply:
Also, looking at the other macro, it's somewhat misleading.  The only flags that are meaningful for library exports are the interrupts enabled / disabled ones.  The registers to clear and stack required are used only by the compartment switcher, which isn't invoked on the libcall path.